haskell/src/Lentille/GraphQL.hs

```diff
@@ -99,71 +104,77 @@ instance From PageInfo Text where
       <> maybe "" show totalCount
       <> (if hasNextPage then " (has next page)" else "")
 
-data RateLimit = RateLimit {used :: Int, remaining :: Int, resetAt :: Text}
+data RateLimit = RateLimit {used :: Int, remaining :: Int, resetAt :: UTCTime}
   deriving (Show)
 
 instance From RateLimit Text where
-  from RateLimit {..} = "remains:" <> show remaining <> ", reset at: " <> resetAt
+  from RateLimit {..} = "remains:" <> show remaining <> ", reset at: " <> show resetAt
 
 streamFetch ::
   (MonadGraphQLE m, Fetch a, FromJSON a, Show a) =>
   GraphClient ->
   -- | query Args constructor, the function takes a cursor
   (Maybe Text -> Args a) ->
+  -- | an action to get a RateLimit record
+  (Maybe (GraphClient -> m RateLimit)) ->
   -- | query result adapter
   (a -> (PageInfo, Maybe RateLimit, [Text], [b])) ->
   Stream (Of b) m ()
-streamFetch client@GraphClient {..} mkArgs transformResponse = go Nothing
+streamFetch client@GraphClient {..} mkArgs getRateLimitM transformResponse = go Nothing
   where
     log :: MonadLog m => Text -> m ()
     log = mLog . Log Macroscope . LogGraphQL lc
       where
         lc = LogCrawlerContext index crawler
 
-    request pageInfoM waitUntilM = do
-      case waitUntilM of
-        Just waitUntil -> do
-          log $ "Reached Quota limit. Waiting until reset date: " <> show waitUntil
-          holdOnUntil waitUntil
-        Nothing -> pure ()
+    holdOnIfNeeded :: (MonadTime m, MonadLog m) => Maybe RateLimit -> m ()
+    holdOnIfNeeded = \case
+      Nothing -> pure ()
+      Just rl -> do
```

Comment:
We can also leverage `sequence_` to do this implementation:

```suggestion
    holdOnIfNeeded :: (MonadTime m, MonadLog m) => Maybe RateLimit -> m ()
    holdOnIfNeeded = sequence_ . fmap toDelay
      where
        toDelay :: RateLimit -> m ()
        toDelay = when (remaining rl <= 0) $ do
            let resetAtTime = resetAt rl
            log $ "Reached Quota limit. Waiting until reset date: " <> show resetAtTime
            holdOnUntil resetAtTime
```

This works because `Maybe` is an instance of `Foldable`:

```
λ> :t sequence_
sequence_ :: (Foldable t, Monad m) => t (m a) -> m ()

λ> sequence_ (Just (putStrLn "hello"))
hello

λ> sequence_ Nothing
```

Finally, note that `sequence` is `traverse id`, so we could also use traverse:

```
λ> traverse_ id (Just (putStrLn "world"))
world
```

Reply:
+1 this makes sense ! even hlint proposed mapM_ that seems to do the fmap for us.

Reply:
Of course `mapM_` works too, it's beautiful =)